Add node scoring logic for pod preemption

bsalamat opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
Scores possible nodes for preemption and picks one node for preemption based on the given score.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
This is based on PR #50190. Please only review "[Add the logic to pick one node for preemption](https://github.com/kubernetes/kubernetes/commit/5141416d9d33d3d6619f9596d112b143f772936e)" here. If you'd like to review other pieces, please go to #50190.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```
ref/ #47604
ref/ #48646

@kubernetes/sig-scheduling-pr-reviews @davidopp

bsalamat committed:
Add preemption victim selector logic to scheduler

bsalamat committed:
autogenerated files

bsalamat committed:
Add the logic to pick one node for preemption

davidopp commented on plugin/pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -445,6 +447,67 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
+	type nodeScore struct {
+		highestPriority int32
+		sumPriorities   int64
+		numPods         int
```
maybe make this int32 since minNumPods := math.MaxInt32 later

bsalamat replied:
math.MaxInt32 is, unlike its name, of type int. We also assign len() to numPods which is also of type int. That's why numPods is an int.

davidopp commented on plugin/pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -445,6 +447,67 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
 }
 
+// pickOneNodeForPreemption chooses one node among the given nodes. It assumes
+// pods in each map entry are ordered by their priority.
```
s/their priority/decreasing priority/

bsalamat replied:
Done

davidopp reviewed (commented):
"Add the logic to pick one node for preemption" commit looks good, just had a couple of tiny comments.

resouer commented on plugin/pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -424,6 +447,176 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
+// pods in each map entry are ordered by their priority.
+// It picks a node based on the following criteria:
+// 1. A node with minimum highest priority victim is picked.
+// 2. Ties are broken by sum of priorities of all victims.
```
Emm ... mind to explain `Ties` here?

bsalamat replied:
It means when there are two or more nodes which are equal based on the previous metric, we order them by the second metric.

bsalamat commented:
@davidopp I found an issue in the initial algorithm when the set of nodes with lowest high priority victim didn't have any common member with the set of nodes with lowest sum of priorities and the set of nodes with minimum number of pods. I changed the code to handle this scenario and added a test case. There is a new commit for the fix. PTAL.

bsalamat committed:
Add specific types for PredicateMetadata and PredicateMetadataProducer

bsalamat commented:
This PR now includes #50805 and #50190. New changes are in https://github.com/kubernetes/kubernetes/pull/50405/commits/ef53496e77dd0777a80b03604757dc7cf80cbc25.

bsalamat commented:
/retest

bsalamat commented:
All the changes in this PR are in #50949 as well.
